test/box-luatest/gh_7362_net_box_async_ping_test.lua

```diff
@@ -0,0 +1,36 @@
+local net = require('net.box')
+local server = require('test.luatest_helpers.server')
+local t = require('luatest')
+local g = t.group()
+
+local TIMEOUT = 5
+
+g.before_all(function()
+    g.server = server:new({alias = 'server'})
+    g.server:start()
+    g.conn = net.connect(g.server.net_box_uri)
+end)
+
+g.after_test('test_connected_ping', function(g)
+    g.server:stop()
+end)
+
+g.after_all(function()
+    g.conn:close()
+end)
+
+g.test_connected_ping = function()
+    t.assert(g.conn:ping({timeout = TIMEOUT}))
+    local future = g.conn:ping({is_async = true})
+    future:wait_result(TIMEOUT)
+    t.assert(future:is_ready())
+    t.assert_equals(future:result(), true)
+end
+
+g.test_disconnected_ping = function(g)
+    t.assert_not(g.conn:ping())
+    -- if connection was closed before the request is done false will be
+    -- returned immediately, not future object.
+    local future = g.conn:ping({is_async = true})
+    t.assert_equals(future, false)
```

Comment:
This looks confusing. All other methods always return a future if `is_async` is true. I think `ping` should always return a future, as well. If the connection is not established, the future should evaluate to `false`.

Reply:
Yep, this looks confusing for me too. And this is what last iteration of reviews was about. 

> All other methods always return a future if `is_async` is true.

Actually they are not:
```lua
-- instance_1
box.cfg({listen = '127.0.0.1:3301'})
-- time point 1
-- kill instance right now
-- time point 2
```
```lua
-- instance_2
conn = require('net.box').connect('0.0.0.0:3301')
conn:eval('return {nil,5}', {}, {is_async = true}) -- here is future object returned
-- time point 1
-- time point 2
conn:eval('return {nil,5}', {}, {is_async = true}) -- error is thrown
---
- error: Peer closed
...
```

E.g. If the peer was closed before making async request, error will be thrown and no future object will be returned. This is due to the checks done in `luaT_netbox_transport_make_request` before registering the request object. So, if we wanted (but we don't)) to return a future object in case of these error, we would have to create a fictive future object (this was implemented one review iteration ago). In the current code it looks like a big crutch. Let's just let `is_async` in ping to act the same way it's done in all other methods.  However, in case of connection related problem it will return false immediately, no error will be thrown as it's done in other `net.box`'s methods. 

The problem here is actually about the test. As I can see, we cannot make the server not to be closed and not to answer to the ping for such a long time, that is enough to close that server. In that case future object will be returned, and after closing the server it will evaluate to `nil, error`.  However this behavior can be simulated in the terminal: 

```lua
-- instance_1
box.cfg({listen = '127.0.0.1:3301'})
while true do end
-- time point 1
-- kill instance right now
```
```lua
-- instance_2
conn = require('net.box').connect('0.0.0.0:3301')
f = conn:eval('return {nil,5}', {}, {is_async = true}) -- here is future object returned
--time point 1
f:is_ready()
---
- true
...
f:result()
---
- null,
- error: Peer closed
...
```

Reply:
> Let's just let `is_async` in ping to act the same way it's done in all other methods. However, in case of connection related problem it will return false immediately, no error will be thrown as it's done in other `net.box`'s methods.

No. This would make `ping` with `is_async` really difficult to use.

Reply:
I'm not sure how async ping should behave in case the connection isn't established. It should probably return a future that evaluates to false. I think you should ask a product manager (@Mons, @unera).

Anyway, we should strive to make the API easy to use, not simplify the code. If the code turns out to be too complicated, better not implement this feature at all, because it isn't planned.

Reply:
Please either ask product managers (I added you to the PM chat) or drop the second patch.

Reply:
It looks like it's a dead end. @Gerold103, please, say, what you think about this.  Moreover, maybe you have some other comments about PR  as code has slightly changed since approval.

Reply:
Also waiting what @Mons will say about the way async ping should behave

Reply:
I think that `:ping()` in an ideal world should return `nil,err` for connection related problems, throw an exception on usage error like a bad option type, and with `is_async` it should return a future only if the request was sent. Otherwise it returns `nil,err` even with `is_async`. That would be consistent with other methods (in a sense that they also immediately fail without any futures even with `is_async` is the connection is not alive from start).

I think that returning just `false` was always a bad idea, because it doesn't tell you the reason which you could at least put into logs. Throwing an exception also could be an option to be even more consistent with other methods, but it would be too much a breaking change. `nil,err` would be at least compatible with old code like `local ok = conn:ping()` and `if conn:ping() then` - these work equally with `false` and `nil,err` returned. We would only potentially break code samples like following: `local res = {conn:ping()}` and `return conn:ping()`.

Otherwise if we want to keep returning `false`, then `is_async` would return either a future or a boolean `false`. User code to check the result would do something like below:
```Lua
local err
local res = conn:ping()
if not res then
    err = box.error.new(box.error.NO_CONNECTION)
else
    res, err = res:wait_result()
end
...
```
Looks a bit unusual that there is no an error object returned on immediate fail, but not too bad. So I would be fine with this solution too.